File: backend/services/currency_conversion.py
"""
Currency Conversion Service
Uses Fixer.io API for historical exchange rates (Bokføringsloven compliance)
"""

import httpx
from decimal import Decimal
from datetime import date
from typing import Optional
from dataclasses import dataclass
import logging

from config.settings import settings

logger = logging.getLogger(__name__)

FIXER_API_URL = "http://data.fixer.io/api"


@dataclass
class ConversionResult:
    """Result of currency conversion."""
    original_currency: str
    original_amount: Decimal
    nok_amount: Decimal
    exchange_rate: Decimal
    rate_date: date
    success: bool
    error: str | None = None


# Cache for exchange rates (simple in-memory cache)
_rate_cache: dict[str, dict] = {}


async def get_exchange_rate(
    from_currency: str,
    to_currency: str = "NOK",
    rate_date: Optional[date] = None
) -> Optional[Decimal]:
    """
    Get exchange rate from Fixer.io API.

    Uses historical rates for compliance with Bokføringsloven.
    Rates are cached to minimize API calls.

    Args:
        from_currency: Source currency (e.g., "USD", "EUR")
        to_currency: Target currency (default "NOK")
        rate_date: Date for historical rate (default: today)

    Returns:
        Exchange rate or None if not available
    """
    if not settings.fixer_api_key:
        logger.warning("Fixer API key not configured")
        return None

    # Normalize currencies
    from_currency = from_currency.upper()
    to_currency = to_currency.upper()

    # Same currency = rate of 1
    if from_currency == to_currency:
        return Decimal("1")

    # Use today if no date specified
    if rate_date is None:
        rate_date = date.today()

    # Check cache
    cache_key = f"{rate_date.isoformat()}_{from_currency}_{to_currency}"
    if cache_key in _rate_cache:
        return _rate_cache[cache_key]["rate"]

    try:
        # Fixer.io free plan only supports EUR as base
        # So we get EUR -> from_currency and EUR -> to_currency, then calculate
        date_str = rate_date.isoformat()

        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{FIXER_API_URL}/{date_str}",
                params={
                    "access_key": settings.fixer_api_key,
                    "symbols": f"{from_currency},{to_currency}",
                }
            )
            response.raise_for_status()
            data = response.json()

        if not data.get("success"):
            error = data.get("error", {}).get("info", "Unknown error")
            logger.error(f"Fixer API error: {error}")
            return None

        rates = data.get("rates", {})

        if from_currency not in rates or to_currency not in rates:
            logger.error(f"Missing rates for {from_currency} or {to_currency}")
            return None

        # Calculate cross rate: from_currency -> EUR -> to_currency
        # rate = (1 / EUR_to_from) * EUR_to_to
        eur_to_from = Decimal(str(rates[from_currency]))
        eur_to_to = Decimal(str(rates[to_currency]))

        rate = eur_to_to / eur_to_from

        # Cache the result
        _rate_cache[cache_key] = {"rate": rate, "date": rate_date}

        logger.info(f"Exchange rate {from_currency}->{to_currency} on {date_str}: {rate}")
        return rate

    except httpx.HTTPError as e:
        logger.error(f"HTTP error fetching exchange rate: {e}")
        return None
    except Exception as e:
        logger.error(f"Error fetching exchange rate: {e}")
        return None


async def convert_to_nok(
    amount: Decimal,
    from_currency: str,
    rate_date: Optional[date] = None
) -> ConversionResult:
    """
    Convert an amount to NOK using historical exchange rate.

    Args:
        amount: Amount in original currency
        from_currency: Source currency code
        rate_date: Date for historical rate (required for compliance)

    Returns:
        ConversionResult with converted amount and rate details
    """
    from_currency = from_currency.upper()

    # Already in NOK
    if from_currency == "NOK":
        return ConversionResult(
            original_currency="NOK",
            original_amount=amount,
            nok_amount=amount,
            exchange_rate=Decimal("1"),
            rate_date=rate_date or date.today(),
            success=True
        )

    # Get exchange rate
    rate = await get_exchange_rate(from_currency, "NOK", rate_date)

    if rate is None:
        # Fallback: use approximate rates for common currencies
        fallback_rates = {
            "USD": Decimal("10.5"),
            "EUR": Decimal("11.5"),
            "GBP": Decimal("13.5"),
            "SEK": Decimal("1.0"),
            "DKK": Decimal("1.55"),
        }
        rate = fallback_rates.get(from_currency)

        if rate is None:
            return ConversionResult(
                original_currency=from_currency,
                original_amount=amount,
                nok_amount=amount,  # No conversion possible
                exchange_rate=Decimal("1"),
                rate_date=rate_date or date.today(),
                success=False,
                error=f"No exchange rate available for {from_currency}"
            )

        logger.warning(f"Using fallback rate for {from_currency}: {rate}")

    # Convert
    nok_amount = (amount * rate).quantize(Decimal("0.01"))

    return ConversionResult(
        original_currency=from_currency,
        original_amount=amount,
        nok_amount=nok_amount,
        exchange_rate=rate,
        rate_date=rate_date or date.today(),
        success=True
    )


def clear_rate_cache():
    """Clear the exchange rate cache."""
    global _rate_cache
    _rate_cache = {}

File: frontend/lib/currency-conversion.ts
/**
 * Currency Conversion Utility
 *
 * Handles conversion of foreign currencies to NOK for Norwegian bookkeeping.
 * Uses Fixer.io API for accurate historical exchange rates.
 *
 * For Norwegian accounting compliance, we use the exchange rate from the
 * invoice date (bilagsdato) as required by Bokføringsloven.
 *
 * @see https://fixer.io/documentation
 */

// Fixer.io API configuration
const FIXER_API_KEY = process.env.NEXT_PUBLIC_FIXER_API_KEY || "";
const FIXER_BASE_URL = "http://data.fixer.io/api";

// Currency symbols and names
export const CURRENCIES: Record<string, { symbol: string; name: string; nameNo: string }> = {
  NOK: { symbol: "kr", name: "Norwegian Krone", nameNo: "Norske kroner" },
  EUR: { symbol: "€", name: "Euro", nameNo: "Euro" },
  USD: { symbol: "$", name: "US Dollar", nameNo: "Amerikanske dollar" },
  GBP: { symbol: "£", name: "British Pound", nameNo: "Britiske pund" },
  SEK: { symbol: "kr", name: "Swedish Krona", nameNo: "Svenske kroner" },
  DKK: { symbol: "kr", name: "Danish Krone", nameNo: "Danske kroner" },
  CHF: { symbol: "Fr", name: "Swiss Franc", nameNo: "Sveitsiske franc" },
  PLN: { symbol: "zł", name: "Polish Zloty", nameNo: "Polske zloty" },
  CZK: { symbol: "Kč", name: "Czech Koruna", nameNo: "Tsjekkiske koruna" },
  HUF: { symbol: "Ft", name: "Hungarian Forint", nameNo: "Ungarske forint" },
  CAD: { symbol: "$", name: "Canadian Dollar", nameNo: "Kanadiske dollar" },
  AUD: { symbol: "$", name: "Australian Dollar", nameNo: "Australske dollar" },
  JPY: { symbol: "¥", name: "Japanese Yen", nameNo: "Japanske yen" },
  CNY: { symbol: "¥", name: "Chinese Yuan", nameNo: "Kinesiske yuan" },
  INR: { symbol: "₹", name: "Indian Rupee", nameNo: "Indiske rupier" },
};

export interface ExchangeRate {
  from: string;
  to: string;
  rate: number;
  date: string;
  source: string;
}

export interface ConversionResult {
  originalAmount: number;
  originalCurrency: string;
  convertedAmount: number;
  targetCurrency: string;
  exchangeRate: number;
  rateDate: string;
  rateSource: string;
}

// Cache for exchange rates by date (for historical lookups)
const rateCacheByDate: Map<string, {
  rates: Record<string, number>;
  timestamp: number;
}> = new Map();

const CACHE_TTL = 24 * 60 * 60 * 1000; // 24 hours for historical rates

/**
 * Fallback rates (updated January 2025)
 * Only used if Fixer.io API is unavailable
 */
const FALLBACK_RATES_TO_NOK: Record<string, number> = {
  EUR: 11.75,
  USD: 11.25,
  GBP: 14.15,
  SEK: 1.01,
  DKK: 1.58,
  CHF: 12.65,
  PLN: 2.72,
  CZK: 0.46,
  CAD: 7.85,
  AUD: 7.05,
  JPY: 0.072,
  CNY: 1.54,
  INR: 0.133,
  NOK: 1.00,
};

interface FixerResponse {
  success: boolean;
  timestamp: number;
  base: string;
  date: string;
  rates: Record<string, number>;
  error?: {
    code: number;
    type: string;
    info: string;
  };
}

/**
 * Fetch exchange rates from Fixer.io for a specific date
 *
 * @param date - Date in YYYY-MM-DD format (for historical rates)
 */
async function fetchFixerRates(date?: string): Promise<{ rates: Record<string, number>; date: string; source: string }> {
  // Check if API key is configured
  if (!FIXER_API_KEY) {
    console.warn("Fixer.io: NEXT_PUBLIC_FIXER_API_KEY not configured, using fallback rates");
    return {
      rates: FALLBACK_RATES_TO_NOK,
      date: date || new Date().toISOString().split("T")[0],
      source: "Fallback rates (Fixer.io not configured)"
    };
  }

  // Use cached rates if available
  const cacheKey = date || "latest";
  const cached = rateCacheByDate.get(cacheKey);
  if (cached && Date.now() - cached.timestamp < CACHE_TTL) {
    return {
      rates: cached.rates,
      date: cacheKey === "latest" ? new Date().toISOString().split("T")[0] : date!,
      source: "Fixer.io (cached)"
    };
  }

  try {
    // Fixer.io free plan only supports EUR as base, so we fetch EUR-based rates
    // and convert to NOK-based rates
    const endpoint = date ? `${FIXER_BASE_URL}/${date}` : `${FIXER_BASE_URL}/latest`;
    const url = `${endpoint}?access_key=${FIXER_API_KEY}&symbols=NOK,USD,GBP,SEK,DKK,CHF,PLN,CZK,CAD,AUD,JPY,CNY,INR,HUF`;

    const response = await fetch(url);
    const data: FixerResponse = await response.json();

    if (!data.success || !data.rates) {
      console.error("Fixer.io API error:", data.error);
      return {
        rates: FALLBACK_RATES_TO_NOK,
        date: date || new Date().toISOString().split("T")[0],
        source: "Fallback rates (API error)"
      };
    }

    // Convert EUR-based rates to NOK-based rates
    // If EUR/NOK = 11.75, then USD/NOK = (EUR/NOK) / (EUR/USD)
    const eurToNok = data.rates.NOK || 11.75;
    const nokBasedRates: Record<string, number> = { NOK: 1, EUR: eurToNok };

    for (const [currency, eurRate] of Object.entries(data.rates)) {
      if (currency !== "NOK") {
        // Convert: 1 CURRENCY = X EUR, 1 EUR = Y NOK
        // So: 1 CURRENCY = X * Y / X = eurToNok / eurRate NOK
        nokBasedRates[currency] = eurToNok / eurRate;
      }
    }

    // Cache the rates
    rateCacheByDate.set(cacheKey, {
      rates: nokBasedRates,
      timestamp: Date.now()
    });

    return {
      rates: nokBasedRates,
      date: data.date,
      source: "Fixer.io"
    };

  } catch (error) {
    console.error("Fixer.io fetch error:", error);
    return {
      rates: FALLBACK_RATES_TO_NOK,
      date: date || new Date().toISOString().split("T")[0],
      source: "Fallback rates (network error)"
    };
  }
}

/**
 * Get exchange rates for a specific date
 * For Norwegian accounting, use the invoice date for compliance
 *
 * @param date - Optional date in YYYY-MM-DD format for historical rates
 */
export async function getExchangeRates(date?: string): Promise<Record<string, number>> {
  const result = await fetchFixerRates(date);
  return result.rates;
}

/**
 * Convert an amount from one currency to NOK
 *
 * @param amount - Amount to convert
 * @param fromCurrency - Source currency code
 * @param invoiceDate - Optional invoice date for historical rate (YYYY-MM-DD)
 */
export async function convertToNOK(
  amount: number,
  fromCurrency: string,
  invoiceDate?: string
): Promise<ConversionResult> {
  if (fromCurrency === "NOK") {
    return {
      originalAmount: amount,
      originalCurrency: "NOK",
      convertedAmount: amount,
      targetCurrency: "NOK",
      exchangeRate: 1,
      rateDate: invoiceDate || new Date().toISOString().split("T")[0],
      rateSource: "No conversion needed"
    };
  }

  // Use invoice date for historical rate (required for Norwegian accounting)
  const { rates, date, source } = await fetchFixerRates(invoiceDate);
  const rate = rates[fromCurrency];

  if (!rate) {
    console.warn(`Unknown currency: ${fromCurrency}, using 1:1 rate`);
    return {
      originalAmount: amount,
      originalCurrency: fromCurrency,
      convertedAmount: amount,
      targetCurrency: "NOK",
      exchangeRate: 1,
      rateDate: date,
      rateSource: "Unknown currency - manual verification required"
    };
  }

  const convertedAmount = Math.round(amount * rate * 100) / 100;

  return {
    originalAmount: amount,
    originalCurrency: fromCurrency,
    convertedAmount,
    targetCurrency: "NOK",
    exchangeRate: Math.round(rate * 10000) / 10000,
    rateDate: date,
    rateSource: source
  };
}

/**
 * Format currency amount for display
 */
export function formatCurrency(
  amount: number,
  currency: string,
  locale: string = "nb-NO"
): string {
  const currencyInfo = CURRENCIES[currency];

  if (currency === "NOK") {
    return `kr ${amount.toLocaleString(locale, { minimumFractionDigits: 0, maximumFractionDigits: 2 })}`;
  }

  // For foreign currencies, show symbol and amount
  const symbol = currencyInfo?.symbol || currency;
  return `${symbol}${amount.toLocaleString(locale, { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`;
}

/**
 * Get currency display info
 */
export function getCurrencyInfo(currency: string): {
  code: string;
  symbol: string;
  name: string;
  nameNo: string;
} {
  const info = CURRENCIES[currency];
  return {
    code: currency,
    symbol: info?.symbol || currency,
    name: info?.name || currency,
    nameNo: info?.nameNo || currency
  };
}

/**
 * Check if currency requires conversion for Norwegian bookkeeping
 */
export function requiresConversion(currency: string): boolean {
  return currency !== "NOK";
}

/**
 * Format exchange rate for display
 */
export function formatExchangeRate(
  fromCurrency: string,
  rate: number
): string {
  return `1 ${fromCurrency} = ${rate.toFixed(4)} NOK`;
}
